### PR TITLE
Add HistoryPreprocessor for Spot and Perpetual

### PR DESCRIPTION
**Pull request by adamiak**
Resolves L2B-663, L2B-826

Review #263 before reviewing this PR.

### CODE REVIEW

File: packages/backend/src/core/preprocessing/PerpetualHistoryPreprocessor.test.ts [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+              prevBalance: 900_000_000n,
+              price: mockCollateralAsset.price,
+              prevPrice: 234_456n,
+              isCurrent: true,
+              prevHistoryId: 99,
+            },
+            {
+              historyId: 80,
+              stateUpdateId: 1940,
+              blockNumber: 9000_000,
```
Comment (Wendrowiec13) [inline packages/backend/src/core/preprocessing/PerpetualHistoryPreprocessor.test.ts:215]: I think it should be `9_000_000`

File: packages/backend/src/core/preprocessing/HistoryPreprocessor.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    trx: Knex.Transaction,
+    newRecords: Omit<
+      PreprocessedAssetHistoryRecord,
+      'historyId' | 'isCurrent'
+    >[]
+  ) {
+    // NOTICE:
+    // We're using a for loop here, because this call:
+    // await this.preprocessedAssetHistoryRepository.addMany(newRecords, trx)
+    // doesn't respect transaction(!!!!):
```
[the remark this answers is not in the document]
Comment (sz-piotr) [reply packages/backend/src/core/preprocessing/HistoryPreprocessor.ts:71]: It's a bug in knex

File: packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+      "This method doesn't respect trascation! Don't use until reviewed!"
+    )
+    const knex = await this.knex(trx)
+    const ids = await knex('preprocessed_asset_history')
+      .insert(rows)
+      .returning('id')
+    return ids.map((x) => x.id)
+  }
+
+  async findById(
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts:91]: findByHistoryId

File: packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    this.getPrevHistoryIdOfCurrentWithStateUpdateId = this.wrapGet(
+      this.getPrevHistoryIdOfCurrentWithStateUpdateId
+    )
+    this.deleteAll = this.wrapDelete(this.deleteAll)
+
+    /* eslint-enable @typescript-eslint/unbound-method */
+  }
+
+  async add(
+    row: Omit<PreprocessedAssetHistoryRecord, 'historyId'>,
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts:66]: record

File: packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    return results[0]!.id
+  }
+
+  async addMany(
+    rows: Omit<PreprocessedAssetHistoryRow, 'id'>[],
+    trx: Knex.Transaction
+  ) {
+    throw new Error(
+      "This method doesn't respect trascation! Don't use until reviewed!"
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts:82]: trascation -> transaction

File: packages/backend/src/core/preprocessing/HistoryPreprocessor.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  // has StarkKey set to ZERO and all assets empty, so we need to retrieve it
+  // all from history
+  async closePositionOrVault(
+    trx: Knex.Transaction,
+    positionId: bigint,
+    stateUpdate: StateUpdateRecord,
+    assetPrices: Record<string, bigint>
+  ) {
+    const recordsToClose =
+      await this.preprocessedAssetHistoryRepository.getCurrentNonEmptyByPositionOrVaultId(
```
Comment (sz-piotr) [inline packages/backend/src/core/preprocessing/HistoryPreprocessor.ts:34]: Why only not empty?

File: packages/backend/src/core/preprocessing/SpotHistoryPreprocessor.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      stateUpdate.id,
+      trx
+    )
+
+    for (const vault of vaults) {
+      if (vault.starkKey === StarkKey.ZERO) {
+        await this.closePositionOrVault(trx, vault.vaultId, stateUpdate, {})
+      } else {
+        const currentRecord = (
+          await this.preprocessedAssetHistoryRepository.getCurrentByStarkKeyAndAssets(
```
Comment (sz-piotr) [inline packages/backend/src/core/preprocessing/SpotHistoryPreprocessor.ts:36]: Why do we need to get the current one? Do we get duplicated vault updates?

**Review by sz-piotr — CHANGES_REQUESTED** [review]

File: packages/backend/src/config/starkex/StarkexConfig.ts [hunk 8]
```diff
@@ -15,6 +15,10 @@ export interface PerpetualRollupConfig {
     proxy: EthereumAddress
     verifiers: EthereumAddress[]
   }
+  collateralAsset: {
+    assetId: AssetId
+    price: bigint
+  }
```
[the remark this answers is not in the document]
Comment (adamiak) [reply packages/backend/src/config/starkex/StarkexConfig.ts]: Done

File: packages/backend/src/core/preprocessing/PerpetualHistoryPreprocessor.test.ts (hunk 1 above)
Comment (adamiak) [reply packages/backend/src/core/preprocessing/PerpetualHistoryPreprocessor.test.ts]: Fixed

File: packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts (hunk 3 above)
Comment (adamiak) [reply packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts]: Fixed

File: packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts (hunk 5 above)
Comment (adamiak) [reply packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts]: Fixed

File: packages/backend/src/core/preprocessing/HistoryPreprocessor.ts (hunk 6 above)
Comment (adamiak) [reply packages/backend/src/core/preprocessing/HistoryPreprocessor.ts]: There was a reason but I found a way to make things even simpler and we no longer need to fetch non-empty here. Thanks :)

File: packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts (hunk 4 above)
Comment (adamiak) [reply packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts:64]: Can you elaborate? :)

Comment (adamiak) [reply packages/backend/src/peripherals/database/PreprocessedAssetHistoryRepository.ts]: Renamed

**Review by sz-piotr — APPROVED** [review]

**Pull request merged**